Infinite Queries: errors when generating TypeScript declaration files

When trying to generate declaration files for an app that uses infinite queries released in v2.6.0, some errors are thrown.

**Reproduction**
https://codesandbox.io/p/devbox/3skwg4?migrateFrom=yw7kg9 

**Steps to reproduce:**
1. Open the reproduction app
2. Run `pnpm run generate-types` (that script uses `tsup`)

**Expected behavior:**
Declaration files are generated without any errors.

**Current behavior:**
Some errors are thrown when generating declaration files:
```
➜  workspace git:(main) ✗ pnpm run generate-types

> react-typescript@1.0.0 generate-types /project/workspace
> tsup ./lib/index.ts --dts-resolve --dts-only -d dist-lib --format esm,cjs

CLI Building entry: lib/index.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v8.4.0
DTS Build start
src/api.ts(33,16): error TS4023: Exported variable 'useBerriesInfiniteQuery' has or is using name 'InfiniteData' from external module "/project/workspace/node_modules/.pnpm/@reduxjs+toolkit@2.6.0_react-redux@9.2.0_react@18.2.0/node_modules/@reduxjs/toolkit/dist/query/react/index" but cannot be named.
src/api.ts(33,16): error TS4023: Exported variable 'useBerriesInfiniteQuery' has or is using name 'UseInfiniteQuerySubscriptionOptions' from external module "/project/workspace/node_modules/.pnpm/@reduxjs+toolkit@2.6.0_react-redux@9.2.0_react@18.2.0/node_modules/@reduxjs/toolkit/dist/query/react/index" but cannot be named.

Error: error occurred in dts build
    at Worker.<anonymous> (/project/workspace/node_modules/.pnpm/tsup@8.4.0_typescript@5.7.3/node_modules/tsup/dist/index.js:1541:26)
    at Worker.emit (node:events:514:28)
    at MessagePort.<anonymous> (node:internal/worker:263:53)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:807:20)
    at exports.emitMessage (node:internal/per_context/messageport:23:28)
DTS Build error
 ELIFECYCLE  Command failed with exit code 1.
```

Possibly related issues:
- https://github.com/reduxjs/redux-toolkit/issues/4875
- https://github.com/reduxjs/redux-toolkit/issues/4863

Possibly related PRs:
- https://github.com/reduxjs/redux-toolkit/pull/4869
- https://github.com/reduxjs/redux-toolkit/pull/4866

The above issues and PRs might be related, I'm not positive though. If that's the case, sorry for a duplicate and feel free to close my issue.

PS: Thank you for implementing infinite queries, they seem work very well. I appreciate it 🙏🏼

--- a/examples/type-portability/bundler/src/app/services/posts.ts
+++ b/examples/type-portability/bundler/src/app/services/posts.ts
@@ -73,6 +73,16 @@ export const postsApi = apiSlice.injectEndpoints({
     getErrorProne: build.query<{ success: boolean }, void>({
       query: () => 'error-prone',
     }),
+    getInfinitePosts: build.infiniteQuery<PostsResponse, void, number>({
+      queryFn: ({ pageParam = 0 }) => ({
+        data: [],
+      }),
+      infiniteQueryOptions: {
+        initialPageParam: 0,
+        getNextPageParam: (lastPage) =>
+          lastPage.length === 0 ? undefined : lastPage[lastPage.length - 1].id,
+      },
+    }),
   }),
 })
 
@@ -87,6 +97,7 @@ export const {
   useLazyGetErrorProneQuery,
   useLazyGetPostQuery,
   useLazyGetPostsQuery,
+  useGetInfinitePostsInfiniteQuery,
   endpoints,
   enhanceEndpoints,
   injectEndpoints,
@@ -106,6 +117,7 @@ export const {
   getPosts,
   login,
   updatePost,
+  getInfinitePosts,
 } = endpoints
 
 export const {
@@ -181,6 +193,19 @@ export const {
   select: _____select,
   useMutation: _____useMutation,
 } = updatePost
+
+export const {
+  Types: ______Types,
+  initiate: ______initiate,
+  matchFulfilled: ______matchFulfilled,
+  matchPending: ______matchPending,
+  matchRejected: ______matchRejected,
+  name: ______name,
+  select: ______select,
+  useInfiniteQueryState: ______useQueryState,
+  useInfiniteQuery: ______useQuery,
+  useInfiniteQuerySubscription: ______useQuerySubscription,
+} = getInfinitePosts
 
 export const {
   internal_getRTKQSubscriptions,

--- a/examples/type-portability/bundler/src/features/posts/PostsManager.tsx
+++ b/examples/type-portability/bundler/src/features/posts/PostsManager.tsx
@@ -8,6 +8,7 @@ import {
   useGetErrorProneQuery,
   useGetPostsQuery,
   useLoginMutation,
+  useGetInfinitePostsInfiniteQuery,
 } from '../../app/services/posts'
 import { logout, selectIsAuthenticated } from '../auth/authSlice'
 import { PostDetail } from './PostDetail'
@@ -70,6 +71,7 @@ export const PostListItem = ({
 
 export const PostList = () => {
   const { data: posts, isLoading } = useGetPostsQuery()
+  useGetInfinitePostsInfiniteQuery()
   const navigate = useNavigate()
 
   if (isLoading) {

--- a/examples/type-portability/nodenext-cjs/src/app/services/posts.ts
+++ b/examples/type-portability/nodenext-cjs/src/app/services/posts.ts
@@ -76,6 +76,18 @@ namespace postsModule {
       getErrorProne: build.query<{ success: boolean }, void>({
         query: () => 'error-prone',
       }),
+      getInfinitePosts: build.infiniteQuery<PostsResponse, void, number>({
+        queryFn: ({ pageParam = 0 }) => ({
+          data: [],
+        }),
+        infiniteQueryOptions: {
+          initialPageParam: 0,
+          getNextPageParam: (lastPage) =>
+            lastPage.length === 0
+              ? undefined
+              : lastPage[lastPage.length - 1].id,
+        },
+      }),
     }),
   })
 
@@ -90,6 +102,7 @@ namespace postsModule {
     useLazyGetErrorProneQuery,
     useLazyGetPostQuery,
     useLazyGetPostsQuery,
+    useGetInfinitePostsInfiniteQuery,
     endpoints,
     enhanceEndpoints,
     injectEndpoints,
@@ -109,6 +122,7 @@ namespace postsModule {
     getPosts,
     login,
     updatePost,
+    getInfinitePosts,
   } = endpoints
 
   export const {
@@ -184,6 +198,19 @@ namespace postsModule {
     select: _____select,
     useMutation: _____useMutation,
   } = updatePost
+
+  export const {
+    Types: ______Types,
+    initiate: ______initiate,
+    matchFulfilled: ______matchFulfilled,
+    matchPending: ______matchPending,
+    matchRejected: ______matchRejected,
+    name: ______name,
+    select: ______select,
+    useInfiniteQueryState: ______useQueryState,
+    useInfiniteQuery: ______useQuery,
+    useInfiniteQuerySubscription: ______useQuerySubscription,
+  } = getInfinitePosts
 
   export const {
     internal_getRTKQSubscriptions,

--- a/examples/type-portability/nodenext-cjs/src/features/posts/PostsManager.tsx
+++ b/examples/type-portability/nodenext-cjs/src/features/posts/PostsManager.tsx
@@ -18,6 +18,7 @@ import useAddPostMutation = postsModule.useAddPostMutation
 import useGetErrorProneQuery = postsModule.useGetErrorProneQuery
 import useGetPostsQuery = postsModule.useGetPostsQuery
 import useLoginMutation = postsModule.useLoginMutation
+import useGetInfinitePostsInfiniteQuery = postsModule.useGetInfinitePostsInfiniteQuery
 import logout = authSliceModule.logout
 import selectIsAuthenticated = authSliceModule.selectIsAuthenticated
 
@@ -81,6 +82,7 @@ const PostListItem = ({
 
 const PostList = () => {
   const { data: posts, isLoading } = useGetPostsQuery()
+  useGetInfinitePostsInfiniteQuery()
   const navigate = useNavigate()
 
   if (isLoading) {

--- a/examples/type-portability/nodenext-esm/src/app/services/posts.ts
+++ b/examples/type-portability/nodenext-esm/src/app/services/posts.ts
@@ -73,6 +73,16 @@ export const postsApi = apiSlice.injectEndpoints({
     getErrorProne: build.query<{ success: boolean }, void>({
       query: () => 'error-prone',
     }),
+    getInfinitePosts: build.infiniteQuery<PostsResponse, void, number>({
+      queryFn: ({ pageParam = 0 }) => ({
+        data: [],
+      }),
+      infiniteQueryOptions: {
+        initialPageParam: 0,
+        getNextPageParam: (lastPage) =>
+          lastPage.length === 0 ? undefined : lastPage[lastPage.length - 1].id,
+      },
+    }),
   }),
 })
 
@@ -87,6 +97,7 @@ export const {
   useLazyGetErrorProneQuery,
   useLazyGetPostQuery,
   useLazyGetPostsQuery,
+  useGetInfinitePostsInfiniteQuery,
   endpoints,
   enhanceEndpoints,
   injectEndpoints,
@@ -106,6 +117,7 @@ export const {
   getPosts,
   login,
   updatePost,
+  getInfinitePosts,
 } = endpoints
 
 export const {
@@ -181,6 +193,19 @@ export const {
   select: _____select,
   useMutation: _____useMutation,
 } = updatePost
+
+export const {
+  Types: ______Types,
+  initiate: ______initiate,
+  matchFulfilled: ______matchFulfilled,
+  matchPending: ______matchPending,
+  matchRejected: ______matchRejected,
+  name: ______name,
+  select: ______select,
+  useInfiniteQueryState: ______useQueryState,
+  useInfiniteQuery: ______useQuery,
+  useInfiniteQuerySubscription: ______useQuerySubscription,
+} = getInfinitePosts
 
 export const {
   internal_getRTKQSubscriptions,

--- a/examples/type-portability/nodenext-esm/src/features/posts/PostsManager.tsx
+++ b/examples/type-portability/nodenext-esm/src/features/posts/PostsManager.tsx
@@ -6,6 +6,7 @@ import type { Post } from '../../app/services/posts.js'
 import {
   useAddPostMutation,
   useGetErrorProneQuery,
+  useGetInfinitePostsInfiniteQuery,
   useGetPostsQuery,
   useLoginMutation,
 } from '../../app/services/posts.js'
@@ -70,6 +71,7 @@ export const PostListItem = ({
 
 export const PostList = () => {
   const { data: posts, isLoading } = useGetPostsQuery()
+  useGetInfinitePostsInfiniteQuery()
   const navigate = useNavigate()
 
   if (isLoading) {

--- a/packages/toolkit/src/query/core/apiState.ts
+++ b/packages/toolkit/src/query/core/apiState.ts
@@ -60,7 +60,7 @@ export type InfiniteQueryConfigOptions<DataType, PageParam> = {
   maxPages?: number
 }
 
-export interface InfiniteData<DataType, PageParam> {
+export type InfiniteData<DataType, PageParam> = {
   pages: Array<DataType>
   pageParams: Array<PageParam>
 }

--- a/packages/toolkit/src/query/endpointDefinitions.ts
+++ b/packages/toolkit/src/query/endpointDefinitions.ts
@@ -544,14 +544,14 @@ export type QueryDefinition<
 > = BaseEndpointDefinition<QueryArg, BaseQuery, ResultType> &
   QueryExtraOptions<TagTypes, ResultType, QueryArg, BaseQuery, ReducerPath>
 
-export interface InfiniteQueryTypes<
+export type InfiniteQueryTypes<
   QueryArg,
   PageParam,
   BaseQuery extends BaseQueryFn,
   TagTypes extends string,
   ResultType,
   ReducerPath extends string = string,
-> extends BaseEndpointTypes<QueryArg, BaseQuery, ResultType> {
+> = BaseEndpointTypes<QueryArg, BaseQuery, ResultType> & {
   /**
    * The endpoint definition type. To be used with some internal generic types.
    * @example

--- a/packages/toolkit/src/query/react/buildHooks.ts
+++ b/packages/toolkit/src/query/react/buildHooks.ts
@@ -809,9 +809,9 @@ export type TypedLazyInfiniteQueryTrigger<
   >
 >
 
-interface UseInfiniteQuerySubscriptionOptions<
+export type UseInfiniteQuerySubscriptionOptions<
   D extends InfiniteQueryDefinition<any, any, any, any, any>,
-> extends SubscriptionOptions {
+> = SubscriptionOptions & {
   /**
    * Prevents a query from automatically running.
    *

